dbms/src/Flash/Mpp/ReceivedMessageQueue.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Mpp/ReceivedMessageQueue.h>
+
+namespace DB
+{
+namespace FailPoints
+{
+extern const char random_receiver_local_msg_push_failure_failpoint[];
+extern const char random_receiver_sync_msg_push_failure_failpoint[];
+extern const char random_receiver_async_msg_push_failure_failpoint[];
+} // namespace FailPoints
+
+namespace
+{
+void injectFailPointReceiverPushFail(bool & push_succeed [[maybe_unused]], ReceiverMode mode)
+{
+    switch (mode)
+    {
+    case ReceiverMode::Local:
+        fiu_do_on(FailPoints::random_receiver_local_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Sync:
+        fiu_do_on(FailPoints::random_receiver_sync_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Async:
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    default:
+        RUNTIME_ASSERT(false, "Illegal ReceiverMode");
+    }
+}
+} // namespace
+
+template <bool fine_grained_shuffle, bool need_wait>
+std::pair<MPMCQueueResult, ReceivedMessagePtr> ReceivedMessageQueue::pop(size_t stream_id)
+{
+    MPMCQueueResult res;
+    ReceivedMessagePtr recv_msg;
```

Comment:
It seems unnecessary to use `shared_ptr`, how about using `ReceivedMessage` directly?

Reply:
Because the queue stores the shared_ptr, if we don't hold the ptr, it will be released.

Reply:
ok, maybe `unique_ptr` can be used here?

Reply:
But it need to be shared among the read streams for fine grained shuffle.

Reply:
ok